src/darwin/Framework/CHIP/MTRDeviceConnectivityMonitor.mm

```diff
@@ -0,0 +1,257 @@
+        _instanceName = [instanceName copy];
+        _connections = [NSMutableDictionary dictionary];
+    }
+    return self;
+}
+
+- (instancetype)initWithCompressedFabricID:(NSNumber *)compressedFabricID nodeID:(NSNumber *)nodeID
+{
+    char instanceName[chip::Dnssd::kMaxOperationalServiceNameSize];
+    chip::PeerId peerId(static_cast<chip::CompressedFabricId>(compressedFabricID.unsignedLongLongValue), static_cast<chip::NodeId>(nodeID.unsignedLongLongValue));
+    CHIP_ERROR err = chip::Dnssd::MakeInstanceName(instanceName, sizeof(instanceName), peerId);
+    if (err != CHIP_NO_ERROR) {
+        MTR_LOG_ERROR("%@ could not make instance name", self);
+        return nil;
+    }
+
+    return [self initWithInstanceName:[NSString stringWithUTF8String:instanceName]];
+}
+
+- (void)dealloc
+{
+    if (_resolver) {
+        DNSServiceRefDeallocate(_resolver);
+    }
+}
+
+- (NSString *)description
+{
+    return [NSString stringWithFormat:@"<MTRDeviceConnectivityMonitor: %@>", _instanceName];
+}
+
++ (DNSServiceRef)_sharedResolverConnection
+{
+    os_unfair_lock_assert_owner(&sConnectivityMonitorLock);
+
+    if (!sSharedResolverConnection) {
+        DNSServiceErrorType dnsError = DNSServiceCreateConnection(&sSharedResolverConnection);
+        if (dnsError != kDNSServiceErr_NoError) {
+            MTR_LOG_ERROR("MTRDeviceConnectivityMonitor: DNSServiceCreateConnection failed %d", dnsError);
+            return NULL;
+        }
+        sSharedResolverQueue = dispatch_queue_create("MTRDeviceConnectivityMonitor", DISPATCH_QUEUE_SERIAL_WITH_AUTORELEASE_POOL);
+        dnsError = DNSServiceSetDispatchQueue(sSharedResolverConnection, sSharedResolverQueue);
+        if (dnsError != kDNSServiceErr_NoError) {
+            MTR_LOG_ERROR("%@ cannot set dispatch queue on resolve", self);
+            DNSServiceRefDeallocate(sSharedResolverConnection);
+            sSharedResolverConnection = NULL;
+            sSharedResolverQueue = nil;
+            return NULL;
+        }
+    }
+
+    return sSharedResolverConnection;
+}
+
+- (void)_callHandler
+{
+    os_unfair_lock_assert_owner(&sConnectivityMonitorLock);
+    MTRDeviceConnectivityMonitorHandler handlerToCall = self->_monitorHandler;
+    if (handlerToCall) {
+        dispatch_async(self->_handlerQueue, handlerToCall);
+    }
+}
+
+- (void)handleResolvedHostname:(const char *)hostName port:(uint16_t)port error:(DNSServiceErrorType)error
+{
+    std::lock_guard lock(sConnectivityMonitorLock);
+
+    // dns_sd.h: must check and call deallocate if error is kDNSServiceErr_ServiceNotRunning
+    if (error == kDNSServiceErr_ServiceNotRunning) {
+        MTR_LOG_ERROR("%@ disconnected from dns-sd subsystem", self);
+        [self _stopMonitoring];
+        return;
+    }
+
+    // Create a nw_connection to monitor connectivity if the host name is not being monitored yet
+    NSString * hostNameString = [NSString stringWithUTF8String:hostName];
+    if (!_connections[hostNameString]) {
```

Comment:
Worth documenting why we are just using the hostname as the key, when the thing we are monitoring is a (host, port) pair.  It's not very obvious.